src/Cardano/Wallet/CoinSelection.hs

```diff
@@ -0,0 +1,154 @@
+      -- and they wish to trasfer an @exact@ amount (or, for example, the max
+      -- amount).
+
+
+data CoinSelectionOptions = CoinSelectionOptions {
+      csoEstimateFee
+        :: Int
+        -> NonEmpty Coin
+        -> Coin
+        -- ^ A function to estimate the fees.
+    , csoExpenseRegulation
+        :: ExpenseRegulation
+        -- ^ A preference regarding expense regulation
+    , csoDustThreshold
+        :: Coin
+        -- ^ Change addresses below the given threshold will be evicted
+        -- from the created transaction. If you only want to remove change
+        -- outputs equal to 0, set 'csoDustThreshold' to 0.
+    }
+
+
+data CoinSelHardErr =
+    CoinSelHardErrOutputCannotCoverFee Text Text
+    -- ^ Payment to receiver insufficient to cover fee
+    -- We record the original output and the fee it needed to cover.
+    -- Only applicable using 'ReceiverPaysFees' regulation
+  | CoinSelHardErrOutputIsRedeemAddress Text
+    -- ^ Attempt to pay into a redeem-only address
+  | CoinSelHardErrCannotCoverFee
+    -- ^ UTxO exhausted whilst trying to pick inputs to cover remaining fee
+  | CoinSelHardErrMaxInputsReached Text
+    -- ^ When trying to construct a transaction, the max number of allowed
+    -- inputs was reached.
+  | CoinSelHardErrUtxoExhausted Text Text
+    -- ^ UTxO exhausted during input selection
+    -- We record the balance of the UTxO as well as the size of the payment
+    -- we tried to make.
+    -- See also 'CoinSelHardErrCannotCoverFee'
+
+type CoinSelPolicy utxo m a =
+       NonEmpty TxOut
+    -> utxo
+    -> m (Either CoinSelHardErr a)
+
+
+newtype MaximumNumOfInputs =
+    MaximumNumOfInputs { getNum :: Word64 }
+
+data CoinSelFinalResult = CoinSelFinalResult
+    { csrInputs  :: NonEmpty (TxIn, TxOut)
+    -- ^ Picked inputs
+    , csrOutputs :: NonEmpty TxOut
+    -- ^ Picked outputs
+    , csrChange  :: [Coin]
+    -- ^ Resulting changes
+    }
+
+
+-- | Random input selection policy
+random
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
+    -> CoinSelPolicy UTxO m CoinSelFinalResult
+random _ = undefined
+
+-- | Largest-first input selection policy
+largestFirst
+    :: forall m. MonadRandom m
+    => CoinSelectionOptions
+    -> MaximumNumOfInputs
+    -> CoinSelPolicy UTxO m CoinSelFinalResult
+largestFirst _ = undefined
+
+
+----------------------------------------------------------------------------
+--                        Fee related                                     --
+----------------------------------------------------------------------------
+
+newtype Fee = Fee { getFee :: Coin }
+
+estimateCardanoFee
+    :: TxSizeLinear
+    -- ^ linear fee policy
+    -> Int
+    -- ^ number of inputs
+    -> [Word64]
+    -- ^ coin value of each output
+    -> Word64
+estimateCardanoFee _ = undefined
+
+estimateMaxTxInputs
+  :: ByteString
+    -- ^ Maximum size of a transaction
+  -> Word64
+estimateMaxTxInputs _ = undefined
```

Comment:
Not needed now. We can pretty much simplify this at the moment by considering:

```hs
estimateFee :: Tx -> Word64

-- or

estimateFee :: Int -> Word64 -> Word64 -- a record type to disambiguate parameters would be good
```

which we can `mock` for now.

Reply:
done